service/qos/service_level_controller.cc

```diff
@@ -221,6 +222,24 @@ future<> service_level_controller::update_service_levels_from_distributed_data()
     });
 }
 
+void service_level_controller::maybe_start_legacy_update_loop(gms::feature_service& feature_service, lw_shared_ptr<db::config> cfg) {
+    if (!feature_service.raft_service_levels_change) {
+        update_from_distributed_data([cfg] () {
+            return std::chrono::duration_cast<steady_clock_type::duration>(std::chrono::milliseconds(cfg->service_levels_interval()));
+        });
+        _global_controller_db->switch_to_raft_sl_change_listener = feature_service.raft_service_levels_change.when_enabled([this] () {
+            stop_update_from_distributed_data();
+        });
+    }
+}
+
+void service_level_controller::stop_update_from_distributed_data() {
+    if (!_global_controller_db || _global_controller_db->dist_data_update_aborter.abort_requested()) {
```

Comment:
Do I understand correctly that `_global_controller_db != nullptr` if and only if `this_shard_id() == 0`?

If so, then it's not useful to be able to call this function on other shards than shard 0. I think it's better to be intentional and just not let call it on other shards. Let's change it so that it asserts that it's called on shard 0 and mention in the docstring as a precondition.